### PR TITLE
Fix: Raise error when no DNA samples linked to RNA subject_id

### PR DESCRIPTION
**Pull request by peterpru**
## Description
In the latest RNA upload PR, an error message was replaced by a new function that counts the number of linked samples to a subject_ID (needed two: one DNA and one RNA sample). However, it did not take into account whether these were RNA or DNA samples, and therefore an important error check was missed in case both RNA samples were non-tumour. In that case it would count each RNA sample, and assert the correct amount of samples were present, even though there were no DNA samples. 

This fix adds in a new staticmethod that checks that each added sample in a case is a MIP-DNA or BALSAMIC case. This new filter then also outputs only these filtered samples, so that the error check now looks for one sample per subject_ID (one DNA sample).

### Added

- staticmethod _filter_mip_dna_balsamic

### Changed

-

### Fixed

- Raise error if no DNA samples connected to RNA sample.

### How to prepare for test
- [ ] ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] paxa the environment: `paxa`
- [ ] install on stage:
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b fix-rna-upload-to-multiple-samples -a`
- [ ] get the associated DNA files to stage `mkdir -p /home/proj/stage/housekeeper-bundles/onkiwi/2021-12-24/` and `cp -r /home/proj/production/housekeeper-bundles/onkiwi/2021-12-24/ /home/proj/stage/housekeeper-bundles/onkiwi/2021-12-24/`
- [ ] do `cg upload upload-case-to-scout onkiwi`
- [ ] get the associated RNA files to stage `mkdir -p /home/proj/stage/housekeeper-bundles/smoothtuna/2022-09-02/` and `cp -r /home/proj/production/housekeeper-bundles/smoothtuna/2022-09-02/ /home/proj/stage/housekeeper-bundles/smoothtuna/2022-09-02/ and cp /home/proj/production/housekeeper-bundles/smoothtuna/2022-09-02/smoothtuna_report.pdf /home/proj/stage/housekeeper-bundles/smoothtuna/2022-09-02/`

#### Test 1
- [ ] In statusdb-stage, removed subject_ID C000350464 from DNA sample ACC9085A13, so that for this subject_ID there are only RNA samples connected. (total is three RNA samples connected, but one of these is a different customer, which is filtered out). The DNA case onkiwi has a total of three samples, which match the three samples in RNA case smoothtuna.
- [ ] do `cg upload rna-to-scout smoothtuna`
- [ ] Screenshot 2022-10-07 at 12 37 44

#### Test 2
- [ ] add subject_id C000350464 back for DNA sample ACC9085A13
- [ ] do `cg upload rna-to-scout smoothtuna`
Screenshot 2022-10-07 at 13 34 06

#### Test 1
- [ ] check that the upload **fails** with the specific error message that no DNA samples are connected for this subject_id, and nothing is uploaded for the other two 'correct' DNA samples.
- [ ] Take a screenshot and attach or copy/paste the output.
- [ ] image

#### Test 2
- [ ] check that upload **succeeds**, with 1 fusion report.pdf, and three bigwig and bed files.
- [ ] Take a screenshot and attach or copy/paste the output.
- [ ] Screenshot 2022-10-07 at 13 46 36

## Review
- [ ] code approved by
- [ ] tests executed by Peter
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan
- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/meta/upload/scout/uploadscoutapi.py [hunk 1]
```diff
@@ -460,3 +458,13 @@ def _add_dna_cases_to_dna_sample(
...
+
+    @staticmethod
+    def _filter_mip_dna_balsamic(subject_id_samples: Query) -> List[Any]:
+        subject_id_dna_samples: List[Any] = []
+        for sample in subject_id_samples.all():
+            for sample.link in sample.links:
+                case_object: models.Family = sample.link.family
+                if case_object.data_analysis in [Pipeline.MIP_DNA, Pipeline.BALSAMIC]:
+                    subject_id_dna_samples.append(sample)
+        return subject_id_dna_samples
```
Comment (karlnyr) [inline cg/meta/upload/scout/uploadscoutapi.py:470]:
```suggestion
    @staticmethod
    def _get_mip_dna_and_balsamic_samples(subject_id_samples: Query) -> List[models.Sample]:
        """
        Filter a models.Sample query, returning samples connected to either a MIP or BALSAMIC case
        """ 
        samples = []
        for sample in subject_id_samples.all():
            for sample_link in sample.links:                
                if sample_link.family.data_analysis in [Pipeline.MIP_DNA, Pipeline.BALSAMIC]:
                    subject_id_dna_samples.append(sample)
        return samples
```

File: cg/meta/upload/scout/uploadscoutapi.py [hunk 2]
```diff
@@ -460,3 +458,13 @@ def _add_dna_cases_to_dna_sample(
                 rna_dna_sample_case_map[rna_sample.internal_id][dna_sample.name].append(
                     case_object.internal_id
                 )
+
+    @staticmethod
+    def _filter_mip_dna_balsamic(subject_id_samples: Query) -> List[Any]:
```
Comment (karlnyr) [inline cg/meta/upload/scout/uploadscoutapi.py:463]: Make the method name more clear. Also, when we are trying to get specific data name it `def _get_bla_bla`. You can also have another look at the return type - the Query will return `models.Sample` when called

Comment (henrikstranneheim) [inline cg/meta/upload/scout/uploadscoutapi.py:463]: Should this be called: "_get_pipelines_subject_id_dna_samples"

**Review by Vince-janv — COMMENTED** [review]
Nice! I think iterating through links is a bit prone to unexpected bugs, since the relationship is many-to-many. I prefer using the appplication tag of the sample

File: cg/meta/upload/scout/uploadscoutapi.py [hunk 3]
```diff
@@ -454,9 +453,24 @@ def _add_dna_cases_to_dna_sample(
...
+    def _get_mip_dna_and_balsamic_samples(subject_id_samples: List[models.Sample]) -> List[Any]:
+        """Filter a models.Sample query, returning DNA samples connected to MIP or BALSAMIC cases"""
+        subject_id_dna_samples: List[Any] = []
+        for sample in subject_id_samples:
+            for sample.link in sample.links:
+                if sample.link.family.data_analysis in [
+                    Pipeline.MIP_DNA,
+                    Pipeline.BALSAMIC,
+                    Pipeline.BALSAMIC_UMI,
+                ]:
```
[the remark this answers is not in the document]
Comment (peterpru) [reply cg/meta/upload/scout/uploadscoutapi.py:474]: Thanks, changed in next commit. Will run some tests now to see that different scenarios perform as expected.

File: cg/meta/upload/scout/uploadscoutapi.py [hunk 4]
```diff
@@ -431,17 +432,15 @@ def _link_rna_sample_to_dna_sample(
             subject_id=rna_sample.subject_id,
             is_tumour=rna_sample.is_tumour,
         )
-        nr_of_subject_id_dna_samples: int = len(subject_id_samples.all())
-        if nr_of_subject_id_dna_samples != 2:
+        subject_id_dna_samples = self._get_application_prep_category(subject_id_samples.all())
+        nr_of_subject_id_dna_samples: int = len(subject_id_dna_samples)
+        if nr_of_subject_id_dna_samples != 1:
```
Comment (Vince-janv) [inline cg/meta/upload/scout/uploadscoutapi.py:437]: I think you can simplify this
```suggestion
        len(subject_id_dna_samples) != 1:
```

**Review by Vince-janv — APPROVED** [review]
Very nice! 🔥 Quick question: Why is an empty fastq-file added? 😅

**Comment by peterpru** [comment]
> Very nice! 🔥 Quick question: Why is an empty fastq-file added? 😅

Thank you for the review. I have removed the empty fastq file from my branch now :). I will test a few scenarios on stage, and then merge.

**Comment by peterpru** [comment]
Test1: Code behaving as expected, upload fails:
image


Test 2: Code still behaving as expected, upload successful:
image

**Comment by peterpru** [comment]
Successful deployment on stage:
image

Successful deployment in production:
image

**Pull request merged**